### ISSUE TITLE
Menu System and Manager Updates

### ISSUE THREAD

**Issue by n00dles** [GetSimpleCMS/GetSimpleCMS#276]
_Original author: ccagle8 (December 31, 2011 14:42:14)_
1. Allow for multi-level menu structures
2. Allow the user to &quot;add&quot; external links into the mix
3. Allow the user to add in new nav items (pages) from this page
4. Change get_navigation() to accommodate multi-level navigation with an added parameter. 

_Original issue: http://code.google.com/p/get-simple-cms/issues/detail?id=276_

**Comment by tablatronix** [comment]
Pages need to have heir own display heirachy, to make page management easy. This can be the old style parent rank system by default. Or else managing thousands of pages is impossible.

Page selector is needed to add pages to menu.
This needs to be simple to use and thought out and tree based on above. Selection needs to be easy.
I prefer a large scrollable combo box or a popup over a select box or a drag combo box, could be an in page tree with a filter, maybe implement something like pages.php in a frame or sorts as it already provides all this functionality.

In page add to menu also needs to be possible , which would pull up menu manager in a popup with the page ready for insertion adjustments.


Menu item extras hooks , plugin need to add custom items, classes rel, icon, etc.

Not sure of external urls should be in core or not.

Not sure if separators should be possible, seperators could also be external links or unlinked classed elements, like parents. This is probably worth putting in if we can make it all in one. Else plugin ready.

**Comment by tablatronix** [comment]
Work on this is being done on a feature branch
https://github.com/GetSimpleCMS/GetSimpleCMS/tree/feature_276_menumanager

I have it working using single page based menus for now, code is all test dev code.
Its going to work pretty well.
I even have default menu modifications updating legacy index/rank in pages

I am going to have to tweak nestable a bit to make it cleaner, fixup padding and make dragging smoother if possible.

No features are added yet, and I am wondering if we should defer this to 3.5

We can add multi level menus to 3.4 then add the extra features to 3.5, this will speed up release of 3.4 by months probably. This feature requires a lot of UI work.

Full feature set 
_Ordered by priority_
* Front end multilevel menu output
* Nested drag and drop UI to reorder menu pages
* Ability to add and remove pages from menu
* Add item for arbitrary url links
* Add items for separators/placeholders or empty parents
* Multiple menus support, with default selection
* Popup menu manager for pages

**Comment by tablatronix** [comment]
I am looking at storing menus as nested arrays with adjacency information, this way we can recurse them extremely fast, and also use them as maps.

something like this


```php
'NESTED' =
array (
  'id' => 'index',
  'data' =>  array (
    'url' => '/dev/getsimple/develop/',
    'path' => 'index',
    'depth' => 1,
    'index' => 9,
    'order' => 7,
  ),
  'numchildren' => 1,
  'children' => array (
      'index-2' => array (
        'id' => 'index-2',
        'parent' => 'index',
        'data' => array (
          'url' => '/dev/getsimple/develop/index/index-2/',
          'path' => 'index/index-2',
          'depth' => 2,
          'index' => 10,
          'order' => 1,
        ),
        'numchildren' => 0,
      ),
    ),  
);

'INDEX' =
array(
  'index' => '.index'
  'index-2' => '.index.index-2'
);

'FLAT' =
array(
  'index' => &['NESTED']['index']['children'],
  'index' => &['[NESTED']['index']['children']['index-2']
);
```

Each menu will be its own array, menus can be stored inside larger arrays of many menus.
Menus can be serialized or stored as json, serialized is not very portable and its slower than json encoding/decoding but it supports references.

The menus NESTED tree can be recursively iterated with no need to do loop lookups.
It also contains adjacency info, parent, level, order, even cached urls and paths etc.

So I am thinking of implementing this using json, with an index map, and rebuilding FLAT arrays from INDEX at runtime, its fairy fast, but it may not be required, and can even be done on demand if required. For small menus it might be optimized to store both in the json, for large this becomes wasteful and references will save load time and memory. tradeoff will have to be determined.

The benefit of having a FLAT array is that we can use it like we do pages now, providing the speed of hash lookups always.

`['FLAT']['slug']['numchildren']`
`['FLAT']['slug']['children']`

**Comment by tablatronix** [comment]
Brain dump , I have been having issues consolidating the feature set for this and am writing is down to try to make sense of it.

Ok well i got all that engineered, but I still have no idea how to implement this.

I was thinking it would be easy, but there is just too much interaction.

We have 3 things that have historically been intertwined, and we are trying to seperate, but I have to keep it simple.

* Page Hierarchy
* Page routes ( fancy urls )
* Menu Hierarchy

These used to be all the same thing, and since adding multilevel paths, and page tree, are now a problem.

We want pages list to have a hierarchy so that its easy to organize/filter/partition pages off however you see fit.

But this does not necessarily need to match menu structure, in fact it cannot or else menu manager needs to contain all pages , not just menu pages to do multi level sorting.

We also have code all over that uses parents, but not always for menus, sometimes for menu dropdowns or selects, so how do we keep this relative for the different uses when we have menu parents and pages parents that are disconnected.

We can have separate `getmenuparent` functions that specify menuids, but this will create backwards functionality issues as all legacy plugins will use the old `pages['parent']` still.

I see no way to solve this unless implementing all features.

Menu manager may have to have a way to organize all pages regardless of `menustatus`.

I do not want to add sorting to pages view, just think it would be too much to have going on.

We could just make the menu separate on upgrade, and from then on its independent from pages hierarchy. I wonder if this will make this less simple though or confusing as it can diverge from pages.

It could be a mode we select, advanced menus or simple menus ( which remain linked to pages hierarchy ) Simple menus could allow menus re-ordering but not nesting changes... but this just seems more confusing in a way.

Basically the only way to allow simple menus and allow multi level is to show all pages in menu manager, and flag or indicate the menu items. But this still creates issues with hundreds of pages.

There is alot of stuff to solve , and maybe I am overcomplicating this, but what happens when you want to add an item in menu from pages, on in page. With advanced menus, we cannot auto add it. We can only auto add it to the simple menu thats linked to pages hierarchy and dynamic.

**Comment by HelgeSverre** [comment]
So is external links in the menu in the 3.4.0 milestone?
I see that it's a highly requested feature all around and none of the plugin solutions work "well".

It has my vote to put it in core.

**Comment by tablatronix** [comment]
yes, but I got stuck on the engineering of just menus..
I am not sure if a UI for it will make it in, but at least some kind of scaffolding will for sure.

**Comment by tablatronix** [comment]
The lack of a framework makes laying out form and ui elements a pita.

**Comment by HelgeSverre** [comment]
Taking some points from the i18n could help, the way the menu is handled in that plugin is very easy and quick to work with.

Maybe creating a UI framework for GetSimple is something that could be looked into, or even adapt something like foundation or bootstrap, and put custom styling over it, it will add a little bloat of course, not sure what the best course of action would be.

**Comment by tablatronix** [comment]
i18n doen'st have external urls.

**Comment by HelgeSverre** [comment]
I was talking about the way you manage the menu hierarchy.

**Comment by tablatronix** [comment]
The way I am managing is described above and already engineered, i am working on the remaining tougher issues.

I want to allow arbitrary urls or placeholders or whatever you want to stick in the menu.
maybe even allowing custom classes and attribs, for separators or whatever and advanced user might want, say menu icons for example, or background colors; and implementing it as simply as possible, and fast on front end.

I also want to allow multiple menus, which may or may not have a gui in core.

I also need a standardized page selector in core, the select dropdown might work for now, but for 500 page users it does not.

I also have to make sure this drag and drop interface can be used on mobiles, or offer a fallback nudge controls. Which might be wise regardless.

Then there is the issue of legacy backwards comparability and how that works with multiple menus, or rather how it does not.

**Comment by tablatronix** [comment]
heres an earlier rough concept and a more practical newer one i am working on now.

![gs_menu_manager_flat](https://cloud.githubusercontent.com/assets/807787/8468240/529283f2-202f-11e5-945e-9f59aabff5cb.png)

![gs_menu_maanger](https://cloud.githubusercontent.com/assets/807787/8468248/87331a5e-202f-11e5-9d35-f7c1ff49b371.png)

**Comment by tablatronix** [comment]
It looks like gs sorts menu order by priority, and blank priority is the highest. We will have to make sure we keep this the same when importing old menus so we don't break menus. Ill have to check and see if that was by design or if this is a new bug.

**Comment by tablatronix** [comment]
This is more an issue with #655 , but switching to multilevel menus will change the default order of menus, so there will have to be a process for switching or a notice.

Old menus are strictly sorted by menuorder, but when we convert it to multi level it will sort by parent hierarchy.

I manual process to create the multi menu, and a warning should suffice.

here is an example of the differences.


![menu_manager_ _page_management_ _gs_develop_-_goog_2015-07-10_08-29-49](https://cloud.githubusercontent.com/assets/807787/8621025/f502e5f4-26e8-11e5-8811-028c835caf18.png)

**Comment by tablatronix** [comment]
Designing the tree iterators now.

I will include several in core probably they will all use callouts with the same structure so they are portable.

The basics are
 * tree from pages using parent hashtable
 * tree from menu cache
 * custom tree from menu cache, needed to recaculate adjacency info for submenus

They will all have 3 callouts
* filter callout - to filter this item from menu, return true on filter, ideally filter your input first..
* outer callout - outerwrap callout for nest containers (ul/ol) - open/close flag
* inner callout - inner item callout for item containers (li) - open/close flag

The entire child will be provided as argument to callouts, additional adjacency info will also be provided as other args, useful for index sorting, level or depth indentation, and order of sub ordering of siblings ( same as we do in legacy pages menus )

`itemcallout($child,$index,$depth,$order,$open);`

adjacency will be cached in menu data for the entire menu, but these will recalculate and provide them also to deal with filtering and subkey output in tree iterators.

```

child-1c [30 -2.3]
child-1c-1b [31 -3.1]
child-1c-1b-1a [32 -4.1]
child-1c-1 [33 -3.2]

reset adjacency
child-1c [1 -1.1]
child-1c-1b [2 -2.1]
child-1c-1b-1a [3 -3.1]
child-1c-1 [4 -2.2]
```

**Comment by tablatronix** [comment]
outstanding design issues

- [ ] There is still the issue of parents, what are parents, in what context are people using them and how do we know what they want, getparent of pages hierarchy, menu hierarchy, routes?

I am going to say we keep using pages hierarchy as the standard core hierarchy, menus can be created but will only be for menus. Page hierarchy is still used for url routing regardless of menus.
I can allow menu manager sorting of page hierarchy and save as menu caches, and also backport into pages fields and pagecache. This solves most of our problem i the short term.

It does not solve for large sites with unnavigable structures, these are still best setup as individual page parent selections, but large dropdowns make selection impossible even with augmented search dropdowns. I am gonig to not attempt to address these issues and design for performance and UI capability for <1000 page sites. Then address these issues in the future separately.

**Comment by tablatronix** [comment]
I added the filters I mentioned before, not sure if anyone would need or use them, but ill probably leave them in , they might be handy in displaying orphaned pages or menu items or something.

They will be built into the tree iterators as custom returns on the filter callouts.
eg. `GSMENUFILTERSKIP`

heres an example of how it works

![gsfiltertype](https://cloud.githubusercontent.com/assets/807787/8806125/ba8de60c-2f9a-11e5-9a84-462ce7819fcd.png)

**Comment by tablatronix** [comment]
Still pondering on the corepages heirarchy.

I have 2 options.
* rebuild the entire menu after page cache is built ( solves for page edit, new, delete, parent change )
* add ability to edit menu items in real time

page cache needs to be fixed so that it is not rebuilding all the time, but our hooks do not support this yet, so it will probably be breaking changes. I will look into it.

rebuilding page cache and menu cache after page update from scratch can take up to 2 seconds for 1000 pages. It is also hard to handle slug changes with page cache, and it depends on pages parent child hierarchies, so the entire thing has to be rebuilt always, no optimizations can be done. 


I have written methods to modify menu item structure in real time per item and even handle slug and parent changes and only rebuild nodes and children in place. This is extremely faster, maybe 50ms total. But it is much more complex i need to hook into all changes to do this. 

Ill post the exact flow later.

But same flags as pagecache updaters
 * page edit ( slug changes, parent change for legacy support which will change around probably)
 * new page
 * delete page
 * restore page

I think it will work well, but i have to do some testing and make sure it will work.
I also have a diff sync, to deal with manual page changes during page cache update.

There is also the issue of using pagecache to perform menu specific functions, requiring page cache to be rebuilt before menu items are rebuilt. eg. generate_permalink() etc.

Ideally everything heirarchy related can be changed to use menu cache much more efficiently, so page cache doesnt even need to be used for this stuff, I think..

**Comment by tablatronix** [comment]
Ok well I did not do a good enough job of engineering how to update a nested tree efficiently, the code became way to convoluted, and it appeared clear that any advantage of speed was going to be useless since the code was spaghetti of references and reindexing maps.

So I yanked it all.

I am going to change this to a hybrid system so that inserts and adjustments are easy and very fast, and storing will be expensive, but reading should still be also fast, so i am thinking of inverting the design I have now to a parent hash table, and a generated tree for recursive manipulation.

Instead of references, the nest will just be id refs to either the parent hash table or a separate data table.

Menu Data can probably be combined with the parent hash or remain as a separate indexed table.
Keeping it separate might be beneficial for splicing menus together or inserting external urls and filtering the menu by them easier. Also modifying the parent hash table can be done without retargeting the data when it moves.

So take 2.

I want output functions to remain recursive, i want the nested array to remain cached.

maybe something like this

```php
<?php

// combined
$menu = 
array(
	'parentid' => array(
		'children' => array('child1','child2'),
  		'data' = array(
    		'parent' => 'parentid',
    		'title'  => 'title'
  		)
	)
);

// sperate
$menu = 
'parents' =
	array(
		'parentid' => 
			array(
				'child1',
				'child2'
			),
	)
'slugs' = array(
	'slug' => 
		array(
	 		'parent' => 'parentid',
	   		'title'  => 'title'
		)
);

// nest built array
array(		
'nest' = array(
			'parentid' = 
				array(
					'child1' => array('grandchild1' => array('greatgrandchild1')),
					'child2' => array('grandchild1'),
				)
		)
);

// with children nodes
array(
'nest' = array(
			'parentid' = 
				array( 'children' =>
					'child1' => array('children' => array('grandchild1' => array('children' => array('greatgrandchild1')))),
					'child2' => array('grandchild1'),
				)
		)
);
```

**Comment by tablatronix** [comment]
Separate data is more likely as it makes insertions and deletions easier, and you could always just rebuild the parent hash table and nest if you wanted to even, so tradeoffs become optional.

I know I am reinventing a design pattern here, but basically a db model for flat files

Separate data array is very compatible with legacy pages parent and makes it easier to rebuild stuff but retain custom data. We can also loop data array for simple iterations, we can loop parent hash for faster child access, and we can directly recurse the nest for full hierarchies.

any references across these can be attached on these points as a data field and adjusted fairly easily.

**Comment by tablatronix** [comment]
Ok changing structure back to data sub arrays

so 

item = array(
'id'
'children'
'data' = array()
)

This way ['data'] can be references and remain the same across flat or nested arrays.

**Comment by tablatronix** [comment]
refactored to use combined parent hash tables for menu data.
With nests available as either parent child maps with no data, static data, or reference data.
Nest references can be rebuilt at runtime, or entire nests can be rebuilt at runtime. This will depend on how I prefer to tune this. 

Having a nest with data allows menus to be passed directly into a recursive tree function with no need to pass down the entire menu object. Otherwise a recursive function needs to have access to the menu object and be able to access all parent hashes at root.

Data links also allow you to traverse the tree directly via [parent]['children'][child], which works well with routing for core menus or custom routing tables ( which are basically the same as menus )

Whether the data is static and saved in file or rebuilt using a reference could be tuned or dependant on size and speed of the menu, but neither is set in stone and have easy to use methods for manipulating them.

This also helps with live modification of menus, the parent hash table can be directly edited and the nest rebuilt. If using refs, data changes other than slug change or parent changes do not require any rebuilds.

```php
$flat = array(
    "parent": {
        "id": "parent",
        "data": {
            "parent": "",
            "depth": 2,
            "index": 8,
            "order": 2,
            "dotpath": ".parent",
            "url": "/dev/getsimple/develop/parent/",
            "path": "parent/child",
            "numchildren": 1
        },
        "children": [
            "child"
        ]
    },
    "child":{...}
) 

$nest = array(
    "parent": {
        "id": "parent",
        "children": {
            "child": {
                "id": "child",
                "data": {...} <- &$flat["parent"]["data"]
            }
        }
    }      
)            
```

**Comment by tablatronix** [comment]
I need to still solve for external links, the unfortunate drawback of doing it this way is that a flat array requires indexes, and external links do not have an index..

So the question is how to implement externals, we can make yet another table that contains externals and merge them in by parent... this also allows links to be passed in via array and not exist in the actual menu. But I am not sure this is a good way to do it.

if would be an external table
```
external[parent] = array(
  item => array(
  url =>   whatever
)
)
```
and when building nests and outputers if(external['parent']) insert
if this is done for nests then outputters might not need to be modified.

Otherwise we need an external flag and a custom index for them.

**Comment by tablatronix** [comment]
- [ ] modify nestable to allow data-attributes that carry through post, i think there is a PR for this already.

**Comment by tablatronix** [comment]
- [ ] generate core menus in update.php
- [ ] generate core pages menu only data `GSMENUIDCOREMENU`
- [ ] update core pages menu only when core menu changes
- [ ] replace parent methods to use defaults

**existing methods for replacement**
_check what they are sorting by, pagessorted etc._
get_navigation
get_pages_menu
get_pages_menu_dropdown
navigation.php
get_available_pages
get_link_menu_array

**Comment by tablatronix** [comment]
- [ ] optimize for menu cache only, pagecache load not necessary ( incude title, url etc ) To avoid having to load parge page caches for just outputting menu

**Comment by tablatronix** [comment]
This is deferred

**Comment by cnb** [comment]
Have you considered making this a plugin?

**Comment by tablatronix** [comment]
not really, the problem is backward compatibility with page files, they just is no way to keep both, there are ways to optimize updating page files to sync with menus, but there really is way to much work required to make pages always update and keep in sync, it is easier to just remove all hierarchy data from page files altogether, and remove support for it, which means breaking all plugins. So i am going to be reworking how this is gonna work.  I will make a easy intermediate where it works as it does not, and only uses page files, and live with the fact that it will be slow as balls and create a cache form it.

**Comment by tablatronix** [comment]
The time i spent trying to make this work the same as it does now, I could have rewritten the entire back end. So I had to give up and spend time more wisely, especially since it is more scarce now.

**Comment by cnb** [comment]
Suggestion: warn if unsaved changes.
I see the form has class `.watch`, but it seems it doesn't work for this.
